### PR TITLE
A user can create a new allocated booking for a person

### PR DESCRIPTION
**Pull request by bryangaledxw**
# Changes in this PR

- A user can book a specific bedspace within a property

## Screenshots of UI changes

localhost_3000_properties_52b62e02-8f72-4baf-b3b2-dd09d3ad395d_bedspaces_f1442ee5-baaf-450a-bcb0-057895a16d03_bookings_new

### CODE REVIEW

**Comment by bryangaledxw** [comment]
[A user can create a new 'allocated' booking for a person](https://trello.com/c/KCMRQZEm/532-a-user-can-create-a-new-allocated-booking-for-a-person)

File: server/utils/dateUtils.ts [hunk 1]
```diff
@@ -30,7 +30,7 @@ export class DateFormats {
     if (options.format === 'long') {
       return format(date, 'cccc d MMMM y')
     } else {
-      return format(date, 'dd/LL/y')
+      return format(date, 'd MMM yy')
```
Comment (pezholio) [inline server/utils/dateUtils.ts:33]: This is a good shout - I think it's more in line with GDS standards, but should it be in its own commit?

File: server/utils/validation.ts [hunk 2]
```diff
@@ -72,6 +72,20 @@ export const errorMessage = (field: string, text: string): ErrorMessage => {
...
+  const data = 'data' in error ? error.data : {}
+  const invalidParams = (data['invalid-params'] ? data['invalid-params'] : []) as Array<Record<string, string>>
+
+  invalidParams.push({
+    propertyName: `$.${propertyName}`,
+    errorType,
+  })
+
+  data['invalid-params'] = invalidParams
+  // eslint-disable-next-line dot-notation
```
Comment (pezholio) [inline server/utils/validation.ts:85]: Is there a reason we can't use dot notation here?

File: server/services/bookingService.ts [hunk 3]
```diff
@@ -21,6 +28,18 @@ export default class BookingService {
     return confirmedBooking
   }
 
+  async createForBedspace(token: string, premisesId: string, room: Room, booking: NewBooking): Promise<Booking> {
+    const bookingClient = this.bookingClientFactory(token)
+
+    const confirmedBooking = await bookingClient.create(premisesId, {
+      serviceName: 'temporary-accommodation',
+      bedId: room.beds[0].id,
```
Comment (pezholio) [inline server/services/bookingService.ts:36]: What happens if a room doesn't have any beds?

File: server/views/temporary-accommodation/bookings/_editable.njk [hunk 4]
```diff
@@ -0,0 +1,54 @@
+{% from "govuk/components/button/macro.njk" import govukButton %}
+{% from "../components/ta-input/macro.njk" import taInput %}
+{% from "../components/ta-date-input/macro.njk" import taDateInput %}
+
+{{ taInput(
+    {
+      label: {
+        text: "What is the person's CRN",
```
Comment (pezholio) [inline server/views/temporary-accommodation/bookings/_editable.njk:8]: I dunno if it might be better from a user's POV if we ask for the person's CRN first, and then confirm it on that page? There might be a risk that this could be booked for the wrong person if the user can't confirm the person's details. Ignore me if this is coming in the future though!

Comment (pezholio) [reply server/views/temporary-accommodation/bookings/_editable.njk:8]: Also, there might be unforseen circumstances if the person is a limited access offender, though I might be getting too deep in the weeds!

File: integration_tests/tests/temporary-accommodation/manage/booking.cy.ts [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    const room = roomFactory.build()
+
+    cy.task('stubSinglePremises', premises)
+    cy.task('stubSingleRoom', { premisesId: premises.id, room })
+
+    // When I visit the new booking page
+    const page = BookingNewPage.visit(premises.id, room.id)
+
+    // And I miss required fields
+    cy.task('stubBookingCreateErrors', {
```
Comment (pezholio) [inline integration_tests/tests/temporary-accommodation/manage/booking.cy.ts:98]: Could we stub a 409 here as well? Just to confirm everything wires up OK

**Review by pezholio — APPROVED** [review]
Looks generally good. Just a couple of things to think about, though you may have already considered them!

File: server/utils/validation.test.ts [hunk 6]
```diff
@@ -180,3 +185,53 @@ describe('fetchErrorsAndUserInput', () => {
     expect(result).toEqual({ errors, errorSummary, userInput })
   })
 })
+
+describe('insertPropertyError', () => {
```
Comment (pezholio) [inline server/utils/validation.test.ts]: Maybe `insertGenericError` would be better, as we already have the concept of `Properties`

File: server/views/temporary-accommodation/bookings/_editable.njk (hunk 4 above)
Comment (bryangaledxw) [reply server/views/temporary-accommodation/bookings/_editable.njk:8]: Yep, we're planning to pull that step out, this is intentionally getting a quick "let us create a booking" in

Comment (pezholio) [reply server/views/temporary-accommodation/bookings/_editable.njk:8]: Cool - I thought that might be the case 👍

File: server/utils/validation.ts (hunk 2 above)
Comment (bryangaledxw) [reply server/utils/validation.ts:85]: Then it fails on the typecheck step, as `data` isn't a known part of the Error type

File: server/utils/dateUtils.ts (hunk 1 above)
Comment (bryangaledxw) [reply server/utils/dateUtils.ts:33]: Done

File: server/utils/validation.test.ts (hunk 6 above)
Comment (bryangaledxw) [reply server/utils/validation.test.ts]: Was torn between `insertPropertyError` and `insertParameterError`, as we use both terms - happy to sidestep the question entirely!

File: server/services/bookingService.ts (hunk 3 above)
Comment (bryangaledxw) [reply server/services/bookingService.ts:36]: I feel ok with this just erroring. All TA rooms should have a bed, and if not, something has gone wrong somewhere. In the case that a room doesn't have a bed, this isn't something we can elegantly handle I feel ok with that erroring here.

Comment (pezholio) [reply server/services/bookingService.ts:36]: 👍

File: integration_tests/tests/temporary-accommodation/manage/booking.cy.ts (hunk 5 above)
Comment (bryangaledxw) [reply integration_tests/tests/temporary-accommodation/manage/booking.cy.ts:98]: Sorted
